[BE] Document optimizer state_dict better, use example

janeyx99 opened the pull request:
![image](https://github.com/pytorch/pytorch/assets/31798555/50ce293c-d884-47ab-b5f5-9ba41e3b4bad)
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105958

janeyx99 committed:
[BE] Document optimizer state_dict better, use example

[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -391,10 +391,39 @@ def state_dict(self):
-            differs between optimizer classes.
-        * param_groups - a list containing all parameter groups where each
-            parameter group is a dict
+        * ``state`` - a Dict holding current optimization state. Its content
```
Do we actually match the id order with the params in the opt we load this into? Or the order of the "state" dict ?

janeyx99 replied:
So it assumes that the two are equivalent and matches the param_groups containing params in the opt to the param ids of the state. For example, if our saved state_dict had:

```
'param_groups': [{ ...,  'params': [0, 1]},  { ..., 'params': [2, 3]}]
```

and our param_groups in the opt we're loading into has:
```
'param_groups': [{ ...,  'params': [tensor_a, tensor_b]},  { ..., 'params': [tensor_c, tensor_d]}]
```

It would just load the state of the 0 id'd param into tensor_a, the 1 id'd param state into tensor_b, and so forth. Does this answer your question?

albanD replied:
My question is really that this doc should explain how we read this state dict if it defines what it contains.
Preferably a single sentence version of this? :D

albanD reviewed (commented):
Very nice!
Small question to add details on how we do order matching.

janeyx99 committed:
Update on "[BE] Document optimizer state_dict better, use example"




[ghstack-poisoned]

janeyx99 committed:
Update on "[BE] Document optimizer state_dict better, use example"


<img width="702" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/aab72afa-2aee-4a7e-809e-78397822df7b">




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good! Thanks for the clarification

janeyx99 committed:
Update on "[BE] Document optimizer state_dict better, use example"

![image](https://github.com/pytorch/pytorch/assets/31798555/0d16fad0-b76d-421e-9023-62a9eae9028e)




[ghstack-poisoned]

janeyx99 committed:
Update on "[BE] Document optimizer state_dict better, use example"

![image](https://github.com/pytorch/pytorch/assets/31798555/0d16fad0-b76d-421e-9023-62a9eae9028e)




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge -i
